orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -691,33 +722,362 @@ Status TransformGraphForPipeline(
+  //    Instead, the input node_arg of nodeD should come from the updated version: updated_node_arg from partition1.
+  //    By using the updated_node_args map, we can retrieve updated_node_arg from original_node_arg, and use that as the
+  //    newly inserted send's input. Also, to keep this on going for any following cut, we create an updated_node_arg_v2,
+  //    and update updated_node_args with updated_node_args[original_node_arg] = updated_node_arg_v2
+  std::map<NodeArg*, NodeArg*> updated_node_args;
+  for (size_t index = 0; index < split_edge_groups_.size(); ++index) {
+    // each entry in split_edge_groups_ represents a partition cut. Each cut can contain the split of
+    // several edges.
+    auto& edgeIds = split_edge_groups_[index];
+
+    // for each cut, record the inserted input/output args.
+    std::vector<NodeArg*> send_input_args;
+    std::vector<NodeArg*> send_output_args;
+    std::vector<NodeArg*> recv_input_args;
+    std::vector<NodeArg*> recv_output_args;
+
+    auto cut_index_str = std::to_string(index);
+    // add input node_arg and initializer for send/recv
+    AddNewNodeArgAndInitializer<bool>(graph,
+                                      "send_input_signal" + cut_index_str,
+                                      ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                      true, /* initializer data */
+                                      send_input_args,
+                                      new_input_names);
+    AddNewNodeArgAndInitializer<bool>(graph,
+                                      "recv_input_signal" + cut_index_str,
+                                      ONNX_NAMESPACE::TensorProto_DataType_BOOL,
+                                      true, /* initializer data */
+                                      recv_input_args,
+                                      new_input_names);
+
+    AddNewNodeArgAndInitializer<size_t>(graph,
+                                      "send_dst_rank" + cut_index_str,
+                                      ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                      index + 1, /* initializer data */
+                                      send_input_args,
+                                      new_input_names);
+    AddNewNodeArgAndInitializer<size_t>(graph,
+                                      "recv_src_rank" + cut_index_str,
+                                      ONNX_NAMESPACE::TensorProto_DataType_INT64,
+                                      index, /* initializer data */
+                                      recv_input_args,
+                                      new_input_names);
+    // add output node_arg for send/recv
+    AddNewNodeArg(graph, "send_output_signal" + cut_index_str, ONNX_NAMESPACE::TensorProto_DataType_BOOL, send_output_args, new_output_names);
+    AddNewNodeArg(graph, "receive_output_signal" + cut_index_str, ONNX_NAMESPACE::TensorProto_DataType_BOOL, recv_output_args, new_output_names);
+
```

Comment:
Just curious. How long a line can be in ORT code base?

Reply:
based on this, it is 120.

[https://github.com/microsoft/onnxruntime/blob/5e0928a777745a1536332973418dbbb7ccbf9d52/docs/Coding_Conventions_and_Standards.md](https://github.com/microsoft/onnxruntime/blob/5e0928a777745a1536332973418dbbb7ccbf9d52/docs/Coding_Conventions_and_Standards.md)


---
In reply to: [429526304](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r429526304) [](ancestors = 429526304)